### PR TITLE
Refactor TrackLogAdminController to support FieldFilter objects for server-side queries

### PR DESCRIPTION
**Pull request by jacob-xhio**
This PR addresses https://github.com/xh/hoist-react/issues/3572

### CODE REVIEW

File: grails-app/controllers/io/xh/hoist/admin/TrackLogAdminController.groovy [hunk 1]
```diff
@@ -11,51 +11,25 @@ import grails.gorm.transactions.ReadOnly
...
-            if (params.browser)     browser =~ "%$params.browser%"
-            if (params.device)      device =~ "%$params.device%"
-            if (params.msg)         msg =~ "%$params.msg%"
-        }
-
-        renderJSON(results)
+        renderJSON(trackLogAdminService.queryTrackLog(parseRequestJSON()))
     }
 
     def lookups() {
```
Comment (lbwexler) [inline grails-app/controllers/io/xh/hoist/admin/TrackLogAdminController.groovy:27]: lookups should come from the service as well

Comment (jacob-xhio) [reply grails-app/controllers/io/xh/hoist/admin/TrackLogAdminController.groovy:26]: Resolved.

File: grails-app/services/io/xh/hoist/track/TrackLogAdminService.groovy [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package io.xh.hoist.track
```
Comment (amcclain) [inline grails-app/services/io/xh/hoist/track/TrackLogAdminService.groovy:1]: Pls add our standard copyright header.  (Does it actually have any magical powers or meaning? 🤷 but we do have it pretty consistently across the source code, and I think it remains a reasonable convention)

File: src/main/groovy/io/xh/hoist/data/filter/CompoundFilter.groovy [hunk 3]
```diff
@@ -0,0 +1,54 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2023 Extremely Heavy Industries Inc.
```
Comment (amcclain) [inline src/main/groovy/io/xh/hoist/data/filter/CompoundFilter.groovy:5]: We can bump these to 2024 as well. It's how I personally keep track of what year it is. 😉

File: src/main/groovy/io/xh/hoist/data/filter/Utils.groovy [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+ *
+ * Copyright © 2023 Extremely Heavy Industries Inc.
+ */
+
+package io.xh.hoist.data.filter
+
+import groovy.transform.CompileStatic
+
+@CompileStatic
+class Utils {
```
Comment (amcclain) [inline src/main/groovy/io/xh/hoist/data/filter/Utils.groovy:13]: Rename as `FilterUtils`?

File: src/main/groovy/io/xh/hoist/data/filter/Utils.groovy [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            }
+        }
+
+        throw new IllegalArgumentException("Unable to identify filter type: ${spec}")
+    }
+
+    /**
+     * Extract a list of fields used by filter
+     */
+    static List getFilterFields(Filter filter) {
```
Comment (amcclain) [inline src/main/groovy/io/xh/hoist/data/filter/Utils.groovy:74]: Return `Set<String>`?

A bit on the fence about letting function filters pass quietly through (vs throwing) - but hard to say what's best without knowing the use case.

**Review by amcclain — COMMENTED** [review]
I know @lbwexler is reviewing so I took a quick look with some minor comments - I admit I didn't carefully look through the actual predicate construction logic as I didn't want to dupe up work, but happy to review more closely if helpful

**Comment by amcclain** [comment]
Will also need changelog entry + determination if it's a breaking change to require https://github.com/xh/hoist-react/pull/3608

File: src/main/groovy/io/xh/hoist/data/filter/Utils.groovy (hunk 5 above)
Comment (jacob-xhio) [reply src/main/groovy/io/xh/hoist/data/filter/Utils.groovy:74]: Will change return type to `Set` and throw runtime exception in function filter case

**Pull request closed without merging**